File: Backend/src/main/java/com/luckyseven/backend/domain/email/service/EmailVerificationService.java
package com.luckyseven.backend.domain.email.service;

import com.luckyseven.backend.domain.email.entity.EmailVerificationToken;
import com.luckyseven.backend.domain.email.mapper.EmailVerificationMapper;
import com.luckyseven.backend.domain.email.repository.EmailVerificationTokenRepository;
import com.luckyseven.backend.domain.member.validate.MemberValidator;
import jakarta.transaction.Transactional;
import java.time.LocalDateTime;
import java.time.temporal.ChronoUnit;
import java.util.UUID;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

@Service
@RequiredArgsConstructor
public class EmailVerificationService {

  private final EmailVerificationTokenRepository tokenRepository;
  private final MemberValidator memberValidator;
  private final EmailVerificationMapper emailVerificationMapper;



  /**
   * 이메일로 인증 토큰 생성 후 DB에 저장
   *
   * @param email 인증 대상 이메일
   * @return 생성된 토큰(UUID)
   */
  @Transactional
  public String generateAndSaveToken(String email){
    tokenRepository.deleteByEmail(email);
    memberValidator.checkDuplicateEmail(email);
    String token = UUID.randomUUID().toString();
    LocalDateTime expiresAt = LocalDateTime.now().plus(1, ChronoUnit.HOURS);
    tokenRepository.save(emailVerificationMapper.makeEntity(email, token, expiresAt));
    return token;
  }


  /**
   * @param token  : token 검증할 토큰(UUID)
   * @return 유효 토큰일 경우 email , 아니라면 null
   */
  public String validateToken(String token){
    return tokenRepository.findByToken(token)
        .filter(t -> t.getExpireTime().isAfter(LocalDateTime.now()))
        .map(EmailVerificationToken::getEmail)
        .orElse(null);
  }


  /**
   * @param email 검증하고나서 검증 성공하면 토큰을 삭제
   */
  @Transactional
  public void deleteTokenByEmail(String email){
    tokenRepository.deleteByEmail(email);
  }

}

File: Backend/src/main/kotlin/com/luckyseven/backend/domain/email/service/EmailVerificationService.kt
package com.luckyseven.backend.domain.email.service

import com.luckyseven.backend.domain.email.entity.EmailVerificationToken
import com.luckyseven.backend.domain.email.repository.EmailVerificationTokenRepository
import jakarta.transaction.Transactional
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service
import java.time.LocalDateTime
import java.util.*

@Service
@Transactional
class EmailVerificationService(
    private val emailVerificationTokenRepository: EmailVerificationTokenRepository
) {
    
    private val logger = LoggerFactory.getLogger(EmailVerificationService::class.java)
    
    fun generateAndSaveToken(email: String): String {
        // 기존 토큰이 있다면 삭제
        emailVerificationTokenRepository.deleteByEmail(email)
        
        // 새 토큰 생성
        val token = UUID.randomUUID().toString()
        val expireTime = LocalDateTime.now().plusHours(24)
        
        val emailToken = EmailVerificationToken(
            token = token,
            email = email,
            expireTime = expireTime
        )
        emailVerificationTokenRepository.save(emailToken)
        
        logger.debug("이메일 인증 토큰 생성: email={}, token={}", email, token.substring(0, 8) + "...")
        return token
    }
    
    fun validateToken(token: String): String? {
        return try {
            val emailToken = emailVerificationTokenRepository.findByToken(token)
                ?: return null
            
            if (emailToken.expireTime.isBefore(LocalDateTime.now())) {
                emailVerificationTokenRepository.delete(emailToken)
                logger.warn("만료된 토큰: {}", token.substring(0, 8) + "...")
                return null
            }
            
            emailToken.email
        } catch (e: Exception) {
            logger.error("토큰 검증 중 오류: {}", token.substring(0, 8) + "...", e)
            null
        }
    }
    
    fun deleteTokenByEmail(email: String) {
        emailVerificationTokenRepository.deleteByEmail(email)
        logger.debug("이메일 토큰 삭제: {}", email)
    }
} 